File: lib/string_finder.cpp
#include "string_finder.hpp"

namespace athes::unpack {
StringFinder::StringFinder(std::shared_ptr<Instruction> ins) : ins(ins) { }

bool StringFinder::is_string() {
    return ins->opcode == OP::getlocal0 && ins->next->opcode == OP::callproperty;
}

bool StringFinder::is_add_string() { return is_string() || ins->opcode == OP::add; }
bool StringFinder::is_next_add_string() {
    return (ins->next->opcode == OP::getlocal0 && ins->next->next->opcode == OP::callproperty)
        || ins->next->opcode == OP::add;
}

bool StringFinder::next_string() {
    while (ins) {
        if (is_string())
            return true;
        ins = ins->next;
    }
    return false;
}

bool StringFinder::next_char(uint32_t& chr) {
    if (!is_add_string())
        return false;

    while (ins->opcode == OP::add)
        ins = ins->next;

    ins = ins->next;
    chr = ins->args[0];
    ins = ins->next;
    return true;
}

void StringFinder::skip_string() {
    while (is_add_string()) {
        if (ins->opcode != OP::add)
            ins = ins->next;

        ins = ins->next;
    }
}

uint32_t StringFinder::addr() { return ins->addr; }

std::string StringFinder::build(std::unordered_map<uint32_t, char> methods) {
    std::string str;
    uint32_t chr;

    if (!is_string())
        next_string();

    while (next_char(chr)) {
        auto it = methods.find(chr);
        if (it != methods.end())
            str.push_back(it->second);
    }

    return str;
}
}

File: src/string_finder.rs
use rabc::abc::parser::{InsIterator, Op, OpCode};
use std::collections::HashMap;

const STRING_SEQUENCE: &[OpCode] = &[OpCode::GetLocal0, OpCode::CallProperty];

pub struct StringFinder<'a> {
    pub prog: &'a mut InsIterator<'a>,
    pub methods: &'a HashMap<u32, u8>,
}

impl<'a> StringFinder<'a> {
    pub fn new(prog: &'a mut InsIterator<'a>, methods: &'a HashMap<u32, u8>) -> Self {
        Self { prog, methods }
    }

    pub fn is_add_string(&self) -> bool {
        self.prog.is(OpCode::Add) || self.prog.is_sequence(STRING_SEQUENCE)
    }

    pub fn match_target(&mut self, target: &str) -> bool {
        for target_byte in target.bytes() {
            if let Some(b) = self.next_char() {
                if self.methods.get(&b) == Some(&target_byte) {
                    continue;
                }
            }
            self.skip_string();
            return false;
        }
        true
    }

    pub fn next_string(&mut self) -> bool {
        self.prog.skip_until_seq(STRING_SEQUENCE).is_some()
    }
    fn skip_string(&mut self) {
        while self.is_add_string() {
            if self.prog.is(OpCode::Add) {
                self.prog.next();
            }
            self.prog.next();
        }
    }

    fn next_char(&mut self) -> Option<u32> {
        if !self.is_add_string() {
            return None;
        }

        while self.prog.is(OpCode::Add) {
            self.prog.next();
        }
        self.prog.next();

        let Op::CallProperty(op) = &self.prog.get().op else {
            return None;
        };
        self.prog.next().and(Some(op.property))
    }

    pub fn build(&mut self) -> Vec<u8> {
        let mut str = Vec::new();
        if !self.prog.is_sequence(STRING_SEQUENCE) {
            self.next_string();
        }

        while let Some(idx) = self.next_char() {
            if let Some(byte) = self.methods.get(&idx) {
                str.push(*byte);
            }
        }
        str
    }
}
